Scripts/Haptics/XRNodeHapticPulse.cs

```diff
@@ -0,0 +1,85 @@
+﻿namespace VRTK.Core.Haptics
+{
+    using UnityEngine;
+    using UnityEngine.XR;
+
+    /// <summary>
+    /// Creates a timed haptic pulse on a <see cref="XRNode"/> supported haptic device.
+    /// </summary>
+    public class XRNodeHapticPulse : HapticPulse
+    {
+        /// <summary>
+        /// The device to pulse.
+        /// </summary>
+        [Tooltip("The device to pulse.")]
+        public XRNode device = XRNode.LeftHand;
+        /// <summary>
+        /// The duration to pulse the device for.
+        /// </summary>
+        [Tooltip("The duration to pulse the device for.")]
+        public float duration = 0.005f;
+
+        /// <summary>
+        /// The haptic capabilities of the specified <see cref="XRNode"/>.
+        /// </summary>
+        protected HapticCapabilities caps = new HapticCapabilities();
+
+        /// <inheritdoc />
+        public override bool IsActive()
+        {
+            return base.IsActive();
+        }
+
+        /// <inheritdoc />
+        protected override void DoBegin()
+        {
+            Pulse(Intensity, duration, device);
+        }
+
+        /// <inheritdoc />
+        protected override void DoCancel()
+        {
+            Pulse(0, 0, device);
+        }
+
+        /// <summary>
+        /// Sends a pulse to the current device.
+        /// </summary>
+        /// <param name="intensity">The intensity to pulse with.</param>
+        /// <param name="duration">The duration to pulse for.</param>
+        /// <param name="node">The device to pulse on.</param>
+        protected virtual void Pulse(float intensity, float duration, XRNode node)
+        {
+            if (!InputDevices.GetDeviceAtXRNode(node).TryGetHapticCapabilities(out caps))
+            {
+                return;
+            }
+
+            if (caps.supportsImpulse)
+            {
+                InputDevices.GetDeviceAtXRNode(node).SendHapticImpulse(0, intensity, duration);
+            }
+            else if (caps.supportsBuffer)
+            {
+                byte[] clip = GeneratePulseBuffer(duration);
+                InputDevices.GetDeviceAtXRNode(node).SendHapticBuffer(0, clip);
+            }
+        }
+
+        /// <summary>
+        /// Generates a pulse buffer array.
+        /// </summary>
+        /// <param name="duration">The duration to pulse for.</param>
+        /// <returns>The buffer array containing the pulse data.</returns>
+        protected virtual byte[] GeneratePulseBuffer(float duration)
+        {
+            int clipCount = (int)(caps.bufferFrequencyHz * duration);
+            byte[] clip = new byte[clipCount];
+            for (int i = 0; i < clipCount; i++)
```

Comment:
Do we do `i` or `index`? I can't remember. 😅

Reply:
looks like `i` in most places and `index` in some.

Personally, in for loops i don't mind a single `i` for the index counter.

But I guess we need to be consistent over the entire code base.

Whatever we decide here, we should apply across the whole codebase soon too.

Reply:
👍 Consistency!

I prefer `index` because if you have nested loops you end up with `i`, `j` and so on which is mainly what I dislike, one also often falls into the trap of doing that even if the index isn't just an index (e.g. math expressions). That's why I settled on always typing it out and then I end up with names like `clipIndex` and `bufferIndex` in the hypothetical example of multiple buffers for multiple clips being generated. Or in math you get reminded to give descriptive names in case the actual index _value_ is of interest - `y` instead of `j` etc.

Reply:
Ok I'll raise a ticket to update everything so it's no longer `i`

Reply:
Feel free to just mark this comment thread as resolved - we can change this `i` to `index` once we do that ticket then.

Reply:
I may as well change this one now as i'm rebasing anyway :)